ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/Basics.hs

```diff
@@ -0,0 +1,104 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Ouroboros.Consensus.Ledger.Tables.Basics (
+    -- * Kinds
+    --
+    -- | For convenience' sake, we define these kinds which convey the intended
+    -- instantiation for the type variables.
+    LedgerStateKind
+  , MapKind
+    -- * Ledger tables
+  , Castable
+  , Key
+  , LedgerTables (..)
+  , Value
+  , castLedgerTables
+  ) where
+
+import           Data.Coerce (coerce)
+import           Data.Kind (Type)
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+import           Ouroboros.Consensus.Ticked (Ticked1)
+
+{-------------------------------------------------------------------------------
+  Kinds
+-------------------------------------------------------------------------------}
+
+-- | Something that holds two types, which intend to represent /keys/ and
+-- /values/.
+type MapKind         = {- key -} Type -> {- value -} Type -> Type
+type LedgerStateKind = MapKind -> Type
+
+{-------------------------------------------------------------------------------
+  Ledger tables
+-------------------------------------------------------------------------------}
+
+-- | The Ledger Tables represent the portion of the data on disk that has been
+-- pulled from disk and attached to the in-memory Ledger State or that will
+-- eventually be written to disk.
+--
+-- With UTxO-HD and the split of the Ledger /ledger state/ into the in-memory
+-- part and the on-disk part, this splitting was reflected in the new type
+-- parameter added to the (Consensus)
+-- 'Ouroboros.Consensus.Ledger.Basics.LedgerState', to which we refer as "the
+-- MapKind" or @mk@.
+--
+-- Every 'Ouroboros.Consensus.Ledger.Basics.LedgerState' (or @LedgerState@-like
+-- type, such as the 'Ouroboros.Consensus.Ledger.Extended.ExtLedgerState') is
+-- associated with a 'LedgerTables' and they both share the @mk@. They both are
+-- of kind 'LedgerStateKind'. 'LedgerTables' is just a way to refer /only/ to a
+-- partial view of the on-disk data without having the rest of the in-memory
+-- 'LedgerState' in scope.
+--
+-- The @mk@ can be instantiated to anything that is map-like, i.e. that expects
+-- two type parameters, the key and the value.
+type LedgerTables :: LedgerStateKind -> MapKind -> Type
+newtype LedgerTables l mk = LedgerTables {
+    getLedgerTables :: mk (Key l) (Value l)
```

Comment:
It used to be a data family `data ledgerState l` that was part of the `HasLedgerTables l` class. I later replaced it with a concrete `LedgerTables` type with `Key`/`Value` type classes. 

Conceptually, this makes sense I think. We assume that the type of data we store is always going to be maplike, i.e., equivalent to `Map k v`. And when we had the data family, each concrete data instance was basically the same except for the UTXO key and value types, and all the boilerplate for defining ledger table combinators was largely the same. Making this a concrete type also made the rest of the code simpler: we define combinators like `lttraverse` only once, and unitary ledger tables (for the most part) nicely compose into hardfork ledger tables with minimal requirements for concrete instantiations of the hard fork block.

I'm not sure what the design is going to look like when we want to store more tables. Presumably these additional tables will stake tables, but we can also assume for the additional tables that they will be map-like. So I would not be surprised if instead of data families we would still be using concrete `LedgerTables` types with some type families to capture the types of keys/values for the tables.